### ISSUE TITLE
Mask decoder finetuning

### ISSUE THREAD

**Issue by thangngoc89** [ChaoningZhang/MobileSAM#49]
Hello and thank you for open sourcing your work. It's truly an amazing engineering feat.
I've tried to fine-tune the mask decoder using **box prompt** with [this repository](https://github.com/luca-medeiros/lightning-sam) as base. It works with VitB, VitL and VitH for my datasets. In all the procedures below, image encoder and prompt encoder layers are always frozen.

However, when I try to fine-tuning the mask decoder of VitT directly, the model seems be falling apart.

Below is the result after 1 epoch vs the original one. Further epoch only shrink the masks further without any visible improvements. 

Result of 1 epoch fine-tuning (VitT)
![image](https://github.com/ChaoningZhang/MobileSAM/assets/3049054/163ebde2-e2ed-4143-9d7b-65390114b9e5)

Original result (VitT)
![image](https://github.com/ChaoningZhang/MobileSAM/assets/3049054/ac26c792-550b-4899-8aca-19c65cb7eaa0)

An exhaustive grid search was performed with 2 hyperparamers: 
- Learning rate: 1e-5, 1e-4, 8e-4, 1e-3, 1e-2
- Batch size: 2, 4, 8, 16, 32, 64, 128, 256 (batch size larger than 32 is trained using gradient accumulation).
In all configuration, the model fell apart after just 1 epoch just like the example above.

Given that VitL of MobileSAM is distilled from VitH while keeping the mask decoder and prompt decoder the same, I've tried another approach:
1. Fine-tune VitH mask decoder
2. Replace the VitL mask decoder with the fine-tuned one from VitH.

This approach worked for me but the results is less than ideal. Some masks are better, some masks are worst comparing to original VitL. Also, fine-tuning VitH takes a lot of computing resources so I would like to avoid that.

Any suggestions to this matter would be much appreciated. 
TIA

**Comment by thangngoc89** [comment]
This is because of incorrectly normalize image during training and inteference. Closing this

**Comment by wm-Githuber** [comment]
> This is because of incorrectly normalize image during training and inteference. Closing this

hello, I have the same problem as you, i.e., I frozen the image encoder and prompt encoder, then fine-tuned the mask decoder. However, I found that the total loss did not decrease with the epoch crease, and the accuracy fluctuated. You said the reason is because of incorrectly normalize image during training and inteference, and I use the original image normalize of SAM. Could you give more detailed information. Thank you very much

**Comment by thangngoc89** [comment]
@wm-Githuber actually in my training setting, the accuracy increased but the original issue is that difference normalization scheme when using MobileSAM. You actually need to not apply normalization.

**Comment by wm-Githuber** [comment]
> @wm-Githuber actually in my training setting, the accuracy increased but the original issue is that difference normalization scheme when using MobileSAM. You actually need to not apply normalization.
Thank you for your reply 👍

**Comment by james-imi** [comment]
@thangngoc89 
were you able to finetune a MobileSAM using this? https://github.com/luca-medeiros/lightning-sam
do you happen to have a code snippet how to do it?
